be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsWithMapHandler.scala

```diff
@@ -0,0 +1,135 @@
+
+        /* building blocks */
+        val handlerPartitioner = builder.add(Partition[GraphMessage](
+          totalPorts,
+          {
+            case Right(jsonRpcMessage: JsonRpcRequest) => jsonRpcMessage.getParams match {
+                case _: Heartbeat       => portHeartBeatHandler
+                case _: GetMessagesById => portGetMessagesByIdHandler
+              }
+            case _ => portPipelineError // Pipeline error goes directly in handlerMerger
+          }
+        ))
+
+        val heartbeatHandler = builder.add(ParamsWithMapHandler.heartbeatHandler(dbActorRef))
+        val getMessagesByIdHandler = builder.add(ParamsWithMapHandler.getMessagesByIdHandler(dbActorRef))
+        val handlerMerger = builder.add(Merge[GraphMessage](totalPorts))
+
+        /* glue the components together */
+        handlerPartitioner.out(portPipelineError) ~> handlerMerger
+        handlerPartitioner.out(portHeartBeatHandler) ~> heartbeatHandler ~> handlerMerger
+        handlerPartitioner.out(portGetMessagesByIdHandler) ~> getMessagesByIdHandler ~> handlerMerger
+
+        /* close the shape */
+        FlowShape(handlerPartitioner.in, handlerMerger.out)
+      }
+  })
+
+  private def heartbeatHandler(dbActorRef: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map({
+    case Right(jsonRpcMessage: JsonRpcRequest) =>
+      /** first step is to retrieve the received heartbeat from the jsonRpcRequest * */
+      val receivedHeartBeat: Map[Channel, Set[Hash]] = jsonRpcMessage.getParams.asInstanceOf[Heartbeat].channelsToMessageIds
+
+      /** second step is to retrieve the local set of channels * */
+      var setOfChannels: Set[Channel] = Set()
+      val ask = dbActorRef ? DbActor.GetAllChannels()
+      Await.ready(ask, duration).value match {
+        case Some(Success(DbActor.DbActorGetAllChannelsAck(channels))) => setOfChannels = channels
+        case Some(Failure(ex: DbActorNAckException))                   => Left(PipelineError(ex.code, s"couldn't retrieve local set of channels", jsonRpcMessage.getId))
+        case reply                                                     => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, s"heartbeatHandler failed : unexpected DbActor reply '$reply'", jsonRpcMessage.getId))
+      }
+
+      /** third step is to ask the DB for the content of each channel in terms of message ids. * */
+      val localHeartBeat: mutable.HashMap[Channel, Set[Hash]] = mutable.HashMap()
+      setOfChannels.foreach(channel => {
+        val ask = dbActorRef ? DbActor.ReadChannelData(channel)
+        Await.ready(ask, duration).value match {
+          case Some(Success(DbActor.DbActorReadChannelDataAck(channelData))) =>
+            val setOfIds = channelData.messages.toSet
+            localHeartBeat += (channel -> setOfIds)
+          case Some(Failure(ex: DbActorNAckException)) => Left(PipelineError(ex.code, s"couldn't readChannelData for local heartbeat", jsonRpcMessage.getId))
```

Comment:
Should we include the specific channel for which it failed in the error message?
```suggestion
          case Some(Failure(ex: DbActorNAckException)) => Left(PipelineError(ex.code, s"ReadChannelData for channel ${channel} local heartbeat", jsonRpcMessage.getId))
```